src/commands/source/download.rs

```diff
@@ -277,39 +297,69 @@ pub async fn download(
                     }
 
                     if source_path_exists && !force {
-                        Err(anyhow!("Source exists: {}", source.path().display()))
+                        info!("Source already exists: {}", source.path().display());
+                        Ok(DownloadResult::Skipped)
                     } else {
+                        let mut download_result: Option<DownloadResult> = None;
+
                         if source_path_exists
                         /* && force is implied by 'if' above*/
                         {
                             source.remove_file().await?;
+                            download_result = Some(DownloadResult::Forced)
                         }
 
                         progressbar.lock().await.inc_download_count().await;
                         {
                             let permit = download_sema.acquire_owned().await?;
                             perform_download(&source, progressbar.clone(), timeout).await?;
                             drop(permit);
+
+                            if download_result.is_none() {
+                                download_result = Some(DownloadResult::Succeeded)
+                            }
                         }
                         progressbar.lock().await.finish_one_download().await;
-                        Ok(())
+                        Ok(download_result.unwrap())
                     }
                 }
             })
         })
         .collect::<futures::stream::FuturesUnordered<_>>()
-        .collect::<Vec<Result<()>>>()
-        .await
-        .into_iter()
-        .collect::<Result<()>>();
+        .collect::<Vec<Result<DownloadResult>>>()
+        .await;
 
-    if r.is_err() {
+    let contains_errors = r.iter().any(|i| i.is_err());
+    if contains_errors {
         progressbar.lock().await.error().await;
-        return r;
     } else {
         progressbar.lock().await.success().await;
     }
 
+    let mut result_types: HashMap<DownloadResult, i32> = HashMap::new();
+    for result in r {
+        match result {
+            Ok(t) => {
+                if let Some(t) = result_types.get_mut(&t) {
+                    *t += 1;
+                } else {
+                    result_types.insert(t, 1);
+                }
+            }
+            Err(_) => {
+                if let Some(t) = result_types.get_mut(&DownloadResult::Error) {
+                    *t += 1;
+                } else {
+                    result_types.insert(DownloadResult::Error, 1);
+                }
+            }
+        }
+    }
+
+    for (r, i) in result_types {
+        println!("{}: {}", i, r.to_verb());
+    }
```

Comment:
That looks super non-idiomatic code to me :o
Might be acceptable though...
I'd prefer a struct with counters but that might be a bit annoying with `async`. We should simplify the datastructures a bit more though by avoiding that `Result<Result>>`.